pkg/secrets/json_expander.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/bitly/go-simplejson"
+)
+
+type JSONExpander struct {
+	sm SecretManager
+}
+
+// NewJSONExpander creates a new secret manager that allows secret values to be stored as json.
+// When resolving secrets, if "dot-notation" is provided via the secret-name, the secret value
+// will be json-decoded and the dot-notation will be used to resolve the secret value.
+func NewJSONExpander(ctx context.Context, f SecretManagerFunc) (SecretManager, error) {
+	sm, err := f(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new JSONExpander SecretManager: %w", err)
+	}
+
+	return WrapJSONExpander(ctx, sm)
+}
+
+// WrapJSONExpander wraps an existing SecretManager with json-expansion logic.
+func WrapJSONExpander(ctx context.Context, sm SecretManager) (SecretManager, error) {
+	return &JSONExpander{
+		sm: sm,
+	}, nil
+}
+
+// GetSecretValue implements the SecretManager interface, but allows for json-expansion
+// of the secret-value. If the secret name contains a period, the secret value is expected
+// to be json. The secret name is assumed to come before the period, while the map-key is
+// expected to follow.
+//
+// For example:
+// If a secret with a name of "psqlcreds" has a value of `{"username":"gandalf", "password":"abc"}`
+// When GetSecretValue(ctx, "psqlcreds") is called, the raw json value will be returned.
+// When GetSecretValue(ctx, "psql.username") is called, only "gandalf" (without quotes) will be returned.
+func (sm *JSONExpander) GetSecretValue(ctx context.Context, name string) (string, error) {
+	parts := strings.Split(name, ".")
+	if len(parts) == 1 {
+		return sm.sm.GetSecretValue(ctx, name)
+	}
+
+	if len(parts) > 2 {
```

Comment:
I think we can handle deep nesting if we use native json:

```golang
var m map[string]interface{}
if err := json.Unmarshal(val, &m); err != nil {
  return err
}

for _, p := range parts[:len(parts)-1] {
  v, ok := m[p]
  if !ok {
    return fmt.Errorf("missing key %v", p)
  }  

  m, ok = v.(map[string]interface{})
  if !ok {
    return fmt.Errorf("not a nested field: %v", p)
  }
}

val, ok := m[len(parts)]
if !ok {
  return fmt.Errorf("missing key %v", p)
}

return val, nil

```

Reply:
I like that idea. I kinda wanted to start with the go-simplejson route just to make sure the general idea is acceptable. I think putting a little work into not bringing in another import makes sense.

Reply:
Done. Thanks for the code. I mostly was able to copy and paste it in place.